### PR TITLE
walker: Use memdb for queued paths

### LINKED ISSUE #838: Many-root workspaces cause server to hang
### Server Version

```
0.26.0
```

### Terraform Version

```
Terraform v1.1.7
on linux_amd64
```

### Client Version

```
Visual Studio Code

Version: 1.65.2
Commit: c722ca6c7eed3d7987c0d5c3df5c45f6b15e77d1
Date: 2022-03-10T15:36:26.048Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Linux x64 5.4.0-104-generic
```

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 4.4.0"
    }
  }
}

resource "aws_s3_bucket" "bucket" {
}
```

### Log Output

https://gist.github.com/chris-redekop/b1e60b409cfa8641e02f54a05aa4a049

### Expected Behavior
The attached [test-3.zip](https://github.com/hashicorp/terraform-ls/files/8326359/test-3.zip) contains a workspace that demonstrates expected behaviour in a 3-root workspace.  After opening this workspace, the VS-Code extension works as expected and I can e.g. format Terraform files.

### Actual Behavior
The attached [test-100.zip](https://github.com/hashicorp/terraform-ls/files/8326360/test-100.zip) contains a workspace that demonstrates problem behaviour in a 100-root workspace.  After opening this workspace, [the terraform-ls logs stop at "watching for changes ..."](https://gist.github.com/chris-redekop/b1e60b409cfa8641e02f54a05aa4a049), and trying to format a Terraform file results in an error: There is no formatter for 'terraform' files installed.

### Steps to Reproduce
Open a many-root workspace, like the one in the attached zip file, and try to format a Terraform file.  Also [a Bash script](https://github.com/hashicorp/terraform-ls/files/8326358/generate-multi-root-workspace.zip) is attached.  It generates a workspace with a specified number of roots.  You might find it useful to create workspaces with different numbers of roots.
```
./generate-multi-root-workspace <n>
```

### PR DESCRIPTION
**Pull request by radeksimko**
Fixes #838 

The bug was originally introduced in https://github.com/hashicorp/terraform-ls/pull/502 - i.e. has been around since introduction of workspace support, for about 9 months. I am therefore leaning towards just slotting this into a regular release schedule (as opposed to cutting patch release).

--- 

Previously we stored queued paths in a priority queue implemented via [`container/heap`](https://pkg.go.dev/container/heap). To allow a more efficient synchronization (which is needed in tests) and a little more readable code (at least the walker code), the data was moved into a new memdb table. This means we can leverage memdb watch channels to do the synchronization and entirely avoid a separate "SyncWalker" implementation just for tests.

## New `walker_paths` memdb table

Each entry within the new table is represented as `WalkerPath`:

https://github.com/hashicorp/terraform-ls/blob/898694f14136985ecdfd78e9b63e50aae77efa9e/internal/state/walker_paths.go#L23-L27

## `WalkerCollector`

A new `WalkerCollector` is introduced to _collect_ job IDs and errors as part of walking, such that we can block & wait for completion of all these async jobs in tests and also fail tests if errors occur during walking.

https://github.com/hashicorp/terraform-ls/blob/898694f14136985ecdfd78e9b63e50aae77efa9e/internal/terraform/module/walker_collector.go#L10-L16

## Synchronization in tests

A new helper function is introduced to block and wait for:

1. walker to finish walking a given path
2. jobs scheduled from that walk to finish

and also to flag up any errors occurred during the walk as test failures.

https://github.com/hashicorp/terraform-ls/blob/898694f14136985ecdfd78e9b63e50aae77efa9e/internal/langserver/handlers/handlers_test.go#L83-L97

This replaces the clunkier sync mode of the walker where `StartWalk()` would optionally block under test.

--- 

## Future - General Purpose Walker?

As noted with the `// TODO` inline comment, I see an opportunity to further cleanup the Walker code and make it more general-purpose by extracting the logic that runs for every module found. That would likely allow us to remove most of the arguments from `NewWalker`. I just didn't want to do it in this PR which already got quite big.

### CODE REVIEW

**Comment by radeksimko** [comment]
Converting to draft as there is still ongoing issue with the channel being filled up due to the fact that we only consume from the channel on a certain occasion.

https://github.com/hashicorp/terraform-ls/blob/bc219ad97f62faa479b98e9092dc5f1b8966f473/internal/terraform/module/walker.go#L191-L204

This would suggest that walker is basically capped 50 paths all the time.

I want to explore the possibility of moving the path queue into memdb as well since the problem we have here is basically the same as in the job scheduler and it might allow us to get rid of most sync logic within the walker.

**Review by dbanck — APPROVED** [review]
👍 thanks for the helpful PR description!

I've only a small suggestion, but that isn't a blocker

File: internal/terraform/module/module_manager_test.go [hunk 1]
```diff
@@ -227,13 +228,31 @@ func TestWalker_complexModules(t *testing.T) {
...
+			dir := document.DirHandleFromPath(tc.root)
+			err = ss.WalkerPaths.EnqueueDir(dir)
+			if err != nil {
+				t.Fatal(err)
+			}
 			err = w.StartWalking(ctx)
 			if err != nil {
 				t.Fatal(err)
 			}
+			err = ss.WalkerPaths.WaitForDirs(ctx, []document.DirHandle{dir})
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply internal/terraform/module/module_manager_test.go:244]: Ah, yes - we can! I just overlooked that opportunity while treating all the test files the same - and the other ones were outside of the `module` package.

[the remark this answers is not in the document]
Comment (radeksimko) [reply internal/terraform/module/module_manager_test.go:244]: Actually that helper method is part of `handlers` so no we can't.

**Pull request merged**